diff --git a/pkg/registry/registrytest/node.go b/pkg/registry/registrytest/node.go
index 9186006a2fc80..35791092d313e 100644
--- a/pkg/registry/registrytest/node.go
+++ b/pkg/registry/registrytest/node.go
@@ -114,5 +114,7 @@ func (r *NodeRegistry) DeleteNode(ctx context.Context, nodeID string) error {
 }
 
 func (r *NodeRegistry) WatchNodes(ctx context.Context, options *metainternalversion.ListOptions) (watch.Interface, error) {
+	r.Lock()
+	defer r.Unlock()
 	return nil, r.Err
 }
