diff --git a/pkg/kv/kvserver/store.go b/pkg/kv/kvserver/store.go
index b8318bf748ff..1f2936447f1d 100644
--- a/pkg/kv/kvserver/store.go
+++ b/pkg/kv/kvserver/store.go
@@ -3443,7 +3443,7 @@ func (s *Store) checkpoint(tag string, spans []roachpb.Span) (string, error) {
 }
 
 // computeMetrics is a common metric computation that is used by
-// ComputeMetricsPeriodically and ComputeMetrics to compute metrics
+// ComputeMetricsPeriodically and ComputeMetrics to compute metrics.
 func (s *Store) computeMetrics(ctx context.Context) (m storage.Metrics, err error) {
 	ctx = s.AnnotateCtx(ctx)
 	if err = s.updateCapacityGauges(ctx); err != nil {
@@ -3473,17 +3473,11 @@ func (s *Store) computeMetrics(ctx context.Context) (m storage.Metrics, err erro
 		s.metrics.RdbCheckpoints.Update(int64(len(dirs)))
 	}
 
-	// Get disk stats for the disk associated with this store.
-	if s.diskMonitor != nil {
-		rollingStats := s.diskMonitor.IncrementalStats()
-		s.metrics.updateDiskStats(rollingStats)
-	}
-
 	return m, nil
 }
 
 // ComputeMetricsPeriodically computes metrics that need to be computed
-// periodically along with the regular metrics
+// periodically along with the regular metrics.
 func (s *Store) ComputeMetricsPeriodically(
 	ctx context.Context, prevMetrics *storage.MetricsForInterval, tick int,
 ) (m storage.Metrics, err error) {
@@ -3491,6 +3485,13 @@ func (s *Store) ComputeMetricsPeriodically(
 	if err != nil {
 		return m, err
 	}
+
+	// Get disk stats for the disk associated with this store.
+	if s.diskMonitor != nil {
+		rollingStats := s.diskMonitor.IncrementalStats()
+		s.metrics.updateDiskStats(rollingStats)
+	}
+
 	wt := m.Flush.WriteThroughput
 
 	if prevMetrics != nil {
diff --git a/pkg/storage/disk/monitor.go b/pkg/storage/disk/monitor.go
index 39103921a91f..2d2ed9f3f0f0 100644
--- a/pkg/storage/disk/monitor.go
+++ b/pkg/storage/disk/monitor.go
@@ -226,8 +226,11 @@ func (s StatsWindow) Latest() Stats {
 type Monitor struct {
 	*monitoredDisk
 
-	// Tracks the time of the last invocation of IncrementalStats.
-	lastIncrementedAt time.Time
+	mu struct {
+		syncutil.Mutex
+		// Tracks the time of the last invocation of IncrementalStats.
+		lastIncrementedAt time.Time
+	}
 }
 
 // CumulativeStats returns the most-recent stats observed.
@@ -241,19 +244,28 @@ func (m *Monitor) CumulativeStats() (Stats, error) {
 	}
 }
 
-// IncrementalStats returns all stats observed since it's previous invocation.
+// updateLastIncrementedAt sets lastIncrementedAt to the current time and
+// returns the previous value.
+func (m *Monitor) swapLastIncrementedAt() time.Time {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	result := m.mu.lastIncrementedAt
+	m.mu.lastIncrementedAt = timeutil.Now()
+	return result
+}
+
+// IncrementalStats returns all stats observed since its previous invocation.
 // Note that the tracer has a bounded capacity and the caller must invoke this
 // method at least as frequently as every COCKROACH_DISK_TRACE_PERIOD to avoid
 // missing events.
 func (m *Monitor) IncrementalStats() StatsWindow {
-	if m.lastIncrementedAt.IsZero() {
-		m.lastIncrementedAt = timeutil.Now()
+	lastIncrementedAt := m.swapLastIncrementedAt()
+	if lastIncrementedAt.IsZero() {
 		return StatsWindow{}
 	}
 
-	events := m.tracer.RollingWindow(m.lastIncrementedAt)
-	m.lastIncrementedAt = timeutil.Now()
-	var stats []Stats
+	events := m.tracer.RollingWindow(lastIncrementedAt)
+	stats := make([]Stats, 0, len(events))
 	for _, event := range events {
 		// Ignore events where we were unable to collect disk stats.
 		if event.err == nil {
diff --git a/pkg/storage/disk/monitor_test.go b/pkg/storage/disk/monitor_test.go
index 10a8382205bf..28a6b74183db 100644
--- a/pkg/storage/disk/monitor_test.go
+++ b/pkg/storage/disk/monitor_test.go
@@ -109,9 +109,9 @@ func TestMonitor_IncrementalStats(t *testing.T) {
 		tracer.RecordEvent(event)
 	}
 	monitor := Monitor{
-		monitoredDisk:     &monitoredDisk{tracer: tracer},
-		lastIncrementedAt: now.Add(-3 * time.Minute),
+		monitoredDisk: &monitoredDisk{tracer: tracer},
 	}
+	monitor.mu.lastIncrementedAt = now.Add(-3 * time.Minute)
 
 	rollingWindow := monitor.IncrementalStats()
 	// Skip the event collected 4 minutes ago since we last incremented 3 minutes ago.
