diff --git a/server/monitor.go b/server/monitor.go
index 6c82a42ad73..00614c506c0 100644
--- a/server/monitor.go
+++ b/server/monitor.go
@@ -320,13 +320,13 @@ func (s *Server) HandleRoutez(w http.ResponseWriter, r *http.Request) {
 			}
 			ri.Subs = castToSliceString(sublist)
 		}
-		r.mu.Unlock()
-
-		if ip, ok := r.nc.(*net.TCPConn); ok {
-			addr := ip.RemoteAddr().(*net.TCPAddr)
+		switch conn := r.nc.(type) {
+		case *net.TCPConn, *tls.Conn:
+			addr := conn.RemoteAddr().(*net.TCPAddr)
 			ri.Port = addr.Port
 			ri.IP = addr.IP.String()
 		}
+		r.mu.Unlock()
 		rs.Routes = append(rs.Routes, ri)
 	}
 	s.mu.Unlock()
