diff --git a/xds/internal/xdsclient/clientimpl_loadreport.go b/xds/internal/xdsclient/clientimpl_loadreport.go
index dd0ae225e8d0..e53df3f1edd9 100644
--- a/xds/internal/xdsclient/clientimpl_loadreport.go
+++ b/xds/internal/xdsclient/clientimpl_loadreport.go
@@ -30,13 +30,15 @@ import (
 func (c *clientImpl) ReportLoad(server *bootstrap.ServerConfig) (*load.Store, func()) {
 	c.authorityMu.Lock()
 	a, err := c.newAuthorityLocked(server)
-	c.authorityMu.Unlock()
 	if err != nil {
+		c.authorityMu.Unlock()
 		c.logger.Infof("xds: failed to connect to the control plane to do load reporting for authority %q: %v", server, err)
 		return nil, func() {}
 	}
 	// Hold the ref before starting load reporting.
 	a.refLocked()
+	c.authorityMu.Unlock()
+
 	store, cancelF := a.reportLoad()
 	return store, func() {
 		cancelF()
