diff --git a/pkg/ccl/serverccl/diagnosticsccl/reporter_test.go b/pkg/ccl/serverccl/diagnosticsccl/reporter_test.go
index d5566c0ac501..86d17b21ea8d 100644
--- a/pkg/ccl/serverccl/diagnosticsccl/reporter_test.go
+++ b/pkg/ccl/serverccl/diagnosticsccl/reporter_test.go
@@ -325,15 +325,15 @@ func TestTelemetry_SuccessfulTelemetryPing(t *testing.T) {
 
 			dr := rt.server.DiagnosticsReporter().(*diagnostics.Reporter)
 
-			before := timeutil.Now()
-			oldTimestamp := dr.LastSuccessfulTelemetryPing
-			require.LessOrEqual(t, dr.LastSuccessfulTelemetryPing, before)
+			before := timeutil.Now().Unix()
+			oldTimestamp := dr.LastSuccessfulTelemetryPing.Load()
+			require.LessOrEqual(t, dr.LastSuccessfulTelemetryPing.Load(), before)
 			dr.ReportDiagnostics(ctx)
 
 			if tc.expectTimestampUpdate {
-				require.GreaterOrEqual(t, dr.LastSuccessfulTelemetryPing, before)
+				require.GreaterOrEqual(t, dr.LastSuccessfulTelemetryPing.Load(), before)
 			} else {
-				require.Equal(t, oldTimestamp, dr.LastSuccessfulTelemetryPing)
+				require.Equal(t, oldTimestamp, dr.LastSuccessfulTelemetryPing.Load())
 			}
 		})
 	}
diff --git a/pkg/server/diagnostics/reporter.go b/pkg/server/diagnostics/reporter.go
index 85a0ecb0d3af..46288fc9685d 100644
--- a/pkg/server/diagnostics/reporter.go
+++ b/pkg/server/diagnostics/reporter.go
@@ -17,6 +17,7 @@ import (
 	"net/http"
 	"net/url"
 	"reflect"
+	"sync/atomic"
 	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/base"
@@ -95,11 +96,12 @@ type Reporter struct {
 	// TestingKnobs is used for internal test controls only.
 	TestingKnobs *TestingKnobs
 
-	// LastSuccessfulTelemetryPing records the current timestamp whenever
-	// we successfully make contact with the registration server. This
-	// timestamp will be updated regardless of whether the response we get
-	// back is successful or not.
-	LastSuccessfulTelemetryPing time.Time
+	// LastSuccessfulTelemetryPing records the current timestamp in
+	// seconds since the Unix epoch whenever we successfully make contact
+	// with the registration server. This timestamp will be updated
+	// regardless of whether the response we get back is successful or
+	// not.
+	LastSuccessfulTelemetryPing atomic.Int64
 }
 
 // PeriodicallyReportDiagnostics starts a background worker that periodically
@@ -176,7 +178,7 @@ func (r *Reporter) ReportDiagnostics(ctx context.Context) {
 	// with the telemetry server and any further problems are not the
 	// customer's fault. We update the telemetry timestamp before moving
 	// on with other request handling.
-	r.LastSuccessfulTelemetryPing = timeutil.Now()
+	r.LastSuccessfulTelemetryPing.Store(timeutil.Now().Unix())
 
 	if res.StatusCode != http.StatusOK {
 		log.Warningf(ctx, "failed to report node usage metrics: status: %s, body: %s", res.Status, b)
