diff --git a/server/filestore.go b/server/filestore.go
index 52f9b32e6bc..21ee18f1bb3 100644
--- a/server/filestore.go
+++ b/server/filestore.go
@@ -5016,7 +5016,10 @@ func (mb *msgBlock) ensurePerSubjectInfoLoaded() error {
 
 // Called on recovery to populate the global psim state.
 func (fs *fileStore) populateGlobalPerSubjectInfo(mb *msgBlock) {
-	if err := mb.readPerSubjectInfo(false); err != nil {
+	mb.mu.Lock()
+	defer mb.mu.Unlock()
+
+	if err := mb.readPerSubjectInfo(true); err != nil {
 		return
 	}
 	// Now populate psim.
