diff --git a/cmd/docker-proxy/udp_proxy_linux.go b/cmd/docker-proxy/udp_proxy_linux.go
index 3796473165dc8..560d893d7ec5a 100644
--- a/cmd/docker-proxy/udp_proxy_linux.go
+++ b/cmd/docker-proxy/udp_proxy_linux.go
@@ -5,6 +5,7 @@ import (
 	"errors"
 	"log"
 	"net"
+	"os"
 	"sync"
 	"syscall"
 	"time"
@@ -14,8 +15,9 @@ import (
 )
 
 const (
-	// UDPConnTrackTimeout is the timeout used for UDP connection tracking
-	UDPConnTrackTimeout = 90 * time.Second
+	// DefaultConnTrackTimeout is the default timeout used for UDP connection
+	// tracking.
+	DefaultConnTrackTimeout = 90 * time.Second
 	// UDPBufSize is the buffer size for the UDP proxy
 	UDPBufSize = 65507
 )
@@ -43,37 +45,65 @@ func newConnTrackKey(addr *net.UDPAddr) *connTrackKey {
 	}
 }
 
-type connTrackMap map[connTrackKey]*net.UDPConn
+type connTrackMap map[connTrackKey]*connTrackEntry
+
+// connTrackEntry wraps a UDP connection to provide thread-safe [net.Conn.Write]
+// and [net.Conn.Close] operations.
+type connTrackEntry struct {
+	conn  *net.UDPConn
+	lastW time.Time
+	// This lock should be held before calling Write or Close on the wrapped
+	// net.UDPConn. Read can be called concurrently to these operations.
+	//
+	// Never lock mu without locking UDPProxy.connTrackLock first.
+	mu sync.Mutex
+}
+
+func newConnTrackEntry(conn *net.UDPConn) *connTrackEntry {
+	return &connTrackEntry{
+		conn: conn,
+		mu:   sync.Mutex{},
+	}
+}
+
+func (cte *connTrackEntry) lastWrite() time.Time {
+	cte.mu.Lock()
+	defer cte.mu.Unlock()
+	return cte.lastW
+}
 
 // UDPProxy is proxy for which handles UDP datagrams. It implements the Proxy
 // interface to handle UDP traffic forwarding between the frontend and backend
 // addresses.
 type UDPProxy struct {
-	listener       *net.UDPConn
-	frontendAddr   *net.UDPAddr
-	backendAddr    *net.UDPAddr
-	connTrackTable connTrackMap
-	connTrackLock  sync.Mutex
-	ipVer          ipVersion
+	listener         *net.UDPConn
+	frontendAddr     *net.UDPAddr
+	backendAddr      *net.UDPAddr
+	connTrackTable   connTrackMap
+	connTrackLock    sync.Mutex
+	connTrackTimeout time.Duration
+	ipVer            ipVersion
 }
 
 // NewUDPProxy creates a new UDPProxy.
 func NewUDPProxy(listener *net.UDPConn, backendAddr *net.UDPAddr, ipVer ipVersion) (*UDPProxy, error) {
 	return &UDPProxy{
-		listener:       listener,
-		frontendAddr:   listener.LocalAddr().(*net.UDPAddr),
-		backendAddr:    backendAddr,
-		connTrackTable: make(connTrackMap),
-		ipVer:          ipVer,
+		listener:         listener,
+		frontendAddr:     listener.LocalAddr().(*net.UDPAddr),
+		backendAddr:      backendAddr,
+		connTrackTable:   make(connTrackMap),
+		connTrackTimeout: DefaultConnTrackTimeout,
+		ipVer:            ipVer,
 	}, nil
 }
 
-func (proxy *UDPProxy) replyLoop(proxyConn *net.UDPConn, serverAddr net.IP, clientAddr *net.UDPAddr, clientKey *connTrackKey) {
+func (proxy *UDPProxy) replyLoop(cte *connTrackEntry, serverAddr net.IP, clientAddr *net.UDPAddr, clientKey *connTrackKey) {
 	defer func() {
 		proxy.connTrackLock.Lock()
 		delete(proxy.connTrackTable, *clientKey)
+		cte.mu.Lock()
 		proxy.connTrackLock.Unlock()
-		proxyConn.Close()
+		cte.conn.Close()
 	}()
 
 	var oob []byte
@@ -87,18 +117,27 @@ func (proxy *UDPProxy) replyLoop(proxyConn *net.UDPConn, serverAddr net.IP, clie
 
 	readBuf := make([]byte, UDPBufSize)
 	for {
-		proxyConn.SetReadDeadline(time.Now().Add(UDPConnTrackTimeout))
+		cte.conn.SetReadDeadline(time.Now().Add(proxy.connTrackTimeout))
 	again:
-		read, err := proxyConn.Read(readBuf)
+		read, err := cte.conn.Read(readBuf)
 		if err != nil {
 			if err, ok := err.(*net.OpError); ok && err.Err == syscall.ECONNREFUSED {
 				// This will happen if the last write failed
 				// (e.g: nothing is actually listening on the
 				// proxied port on the container), ignore it
-				// and continue until UDPConnTrackTimeout
+				// and continue until DefaultConnTrackTimeout
 				// expires:
 				goto again
 			}
+			// If the UDP connection is one-sided (i.e. the backend never sends
+			// replies), the connTrackEntry should not be GC'd until no writes
+			// happen for proxy.connTrackTimeout.
+			//
+			// Since the ReadDeadline is set to proxy.connTrackTimeout, in such
+			// case, the connTrackEntry will be GC'd at most after 2 * proxy.connTrackTimeout.
+			if errors.Is(err, os.ErrDeadlineExceeded) && time.Since(cte.lastWrite()) < proxy.connTrackTimeout {
+				continue
+			}
 			return
 		}
 		for i := 0; i != read; {
@@ -134,15 +173,16 @@ func (proxy *UDPProxy) Run() {
 
 		fromKey := newConnTrackKey(from)
 		proxy.connTrackLock.Lock()
-		proxyConn, hit := proxy.connTrackTable[*fromKey]
+		cte, hit := proxy.connTrackTable[*fromKey]
 		if !hit {
-			proxyConn, err = net.DialUDP("udp", nil, proxy.backendAddr)
+			proxyConn, err := net.DialUDP("udp", nil, proxy.backendAddr)
 			if err != nil {
 				log.Printf("Can't proxy a datagram to udp/%s: %s\n", proxy.backendAddr, err)
 				proxy.connTrackLock.Unlock()
 				continue
 			}
-			proxy.connTrackTable[*fromKey] = proxyConn
+			cte = newConnTrackEntry(proxyConn)
+			proxy.connTrackTable[*fromKey] = cte
 
 			daddr, err := readDestFromCmsg(oob, proxy.ipVer)
 			if err != nil {
@@ -151,17 +191,21 @@ func (proxy *UDPProxy) Run() {
 				continue
 			}
 
-			go proxy.replyLoop(proxyConn, daddr, from, fromKey)
+			go proxy.replyLoop(cte, daddr, from, fromKey)
 		}
+		cte.mu.Lock()
 		proxy.connTrackLock.Unlock()
+		cte.conn.SetWriteDeadline(time.Now().Add(proxy.connTrackTimeout))
 		for i := 0; i != read; {
-			written, err := proxyConn.Write(readBuf[i:read])
+			written, err := cte.conn.Write(readBuf[i:read])
 			if err != nil {
 				log.Printf("Can't proxy a datagram to udp/%s: %s\n", proxy.backendAddr, err)
 				break
 			}
 			i += written
+			cte.lastW = time.Now()
 		}
+		cte.mu.Unlock()
 	}
 }
 
@@ -194,12 +238,15 @@ func readDestFromCmsg(oob []byte, ipVer ipVersion) (_ net.IP, err error) {
 	return cm.Dst, nil
 }
 
-// Close stops forwarding the traffic.
+// Close ungracefully stops forwarding the traffic.
 func (proxy *UDPProxy) Close() {
 	proxy.listener.Close()
 	proxy.connTrackLock.Lock()
 	defer proxy.connTrackLock.Unlock()
-	for _, conn := range proxy.connTrackTable {
-		conn.Close()
+	for _, cte := range proxy.connTrackTable {
+		// Unlike the GC logic in replyLoop, we want to close the connections
+		// immediately, even if there are pending and in-progress writes. So no
+		// need to lock cte.mu here.
+		cte.conn.Close()
 	}
 }
diff --git a/cmd/docker-proxy/udp_proxy_linux_test.go b/cmd/docker-proxy/udp_proxy_linux_test.go
new file mode 100644
index 0000000000000..e2f903d8b28c1
--- /dev/null
+++ b/cmd/docker-proxy/udp_proxy_linux_test.go
@@ -0,0 +1,78 @@
+package main
+
+import (
+	"net"
+	"testing"
+	"time"
+
+	"gotest.tools/v3/assert"
+)
+
+// TestUDPOneSided makes sure that the conntrack entry isn't GC'd if the
+// backend never writes to the UDP client.
+func TestUDPOneSided(t *testing.T) {
+	frontend, err := net.ListenUDP("udp", &net.UDPAddr{IP: net.IPv4(127, 0, 0, 1), Port: 0})
+	assert.NilError(t, err)
+	defer frontend.Close()
+
+	backend, err := net.ListenUDP("udp", &net.UDPAddr{IP: net.IPv4(127, 0, 0, 1), Port: 0})
+	assert.NilError(t, err)
+	defer backend.Close()
+
+	type udpMsg struct {
+		data  []byte
+		saddr *net.UDPAddr
+	}
+	msgs := make(chan udpMsg)
+	go func() {
+		for {
+			buf := make([]byte, 1024)
+			n, saddr, err := backend.ReadFromUDP(buf)
+			if err != nil {
+				return
+			}
+			msgs <- udpMsg{data: buf[:n], saddr: saddr}
+		}
+	}()
+
+	proxy, err := NewUDPProxy(frontend, backend.LocalAddr().(*net.UDPAddr), ip4)
+	assert.NilError(t, err)
+	defer proxy.Close()
+
+	const connTrackTimeout = 1 * time.Second
+	proxy.connTrackTimeout = connTrackTimeout
+
+	go func() {
+		proxy.Run()
+	}()
+
+	client, err := net.DialUDP("udp", nil, frontend.LocalAddr().(*net.UDPAddr))
+	assert.NilError(t, err)
+	defer client.Close()
+
+	var expSaddr *net.UDPAddr
+	for i := range 15 {
+		_, err = client.Write([]byte("hello"))
+		assert.NilError(t, err)
+		time.Sleep(100 * time.Millisecond)
+
+		msg := <-msgs
+		assert.Equal(t, string(msg.data), "hello")
+		if i == 0 {
+			expSaddr = msg.saddr
+		} else {
+			assert.Equal(t, msg.saddr.Port, expSaddr.Port)
+		}
+	}
+
+	// The conntrack entry is checked every connTrackTimeout, but the latest
+	// write might be less than connTrackTimeout ago. So we need to wait for
+	// at least twice the conntrack timeout to make sure the entry is GC'd.
+	time.Sleep(2 * connTrackTimeout)
+	_, err = client.Write([]byte("hello"))
+	assert.NilError(t, err)
+
+	msg := <-msgs
+	assert.Equal(t, string(msg.data), "hello")
+	assert.Check(t, msg.saddr.Port != expSaddr.Port)
+}
