diff --git a/scrape/target.go b/scrape/target.go
index 6241d52281e..d7a97155160 100644
--- a/scrape/target.go
+++ b/scrape/target.go
@@ -110,6 +110,8 @@ func (t *Target) Labels() labels.Labels {
 
 // DiscoveredLabels returns a copy of the target's labels before any processing.
 func (t *Target) DiscoveredLabels() labels.Labels {
+	t.mtx.Lock()
+	defer t.mtx.Unlock()
 	lset := make(labels.Labels, len(t.discoveredLabels))
 	copy(lset, t.discoveredLabels)
 	return lset
@@ -117,6 +119,8 @@ func (t *Target) DiscoveredLabels() labels.Labels {
 
 // SetDiscoveredLabels sets new DiscoveredLabels
 func (t *Target) SetDiscoveredLabels(l labels.Labels) {
+	t.mtx.Lock()
+	defer t.mtx.Unlock()
 	t.discoveredLabels = l
 }
 
