diff --git a/pkg/roachprod/ui/spinner.go b/pkg/roachprod/ui/spinner.go
index 1745bfe06b64..c6a68615d11a 100644
--- a/pkg/roachprod/ui/spinner.go
+++ b/pkg/roachprod/ui/spinner.go
@@ -118,9 +118,9 @@ func (s *Spinner) Start() func() {
 			fmt.Fprintf(s.out, "\n")
 		}
 	}
+	s.waitGroup.Add(1)
 	go func() {
 		defer s.waitGroup.Done()
-		s.waitGroup.Add(1)
 
 		var writer Writer
 		tickerDuration := 100 * time.Millisecond
