diff --git a/internal/grpctest/tlogger.go b/internal/grpctest/tlogger.go
index 23d4a4e42fd1..9abf1acb1e6e 100644
--- a/internal/grpctest/tlogger.go
+++ b/internal/grpctest/tlogger.go
@@ -26,6 +26,7 @@ import (
 	"runtime/debug"
 	"strconv"
 	"strings"
+	"sync"
 	"testing"
 
 	"google.golang.org/grpc/grpclog"
@@ -48,18 +49,21 @@ const (
 type tLogger struct {
 	v           int
 	t           *testing.T
-	errors      map[*regexp.Regexp]int
 	initialized bool
+
+	m      sync.Mutex // protects errors
+	errors map[*regexp.Regexp]int
 }
 
 func init() {
-	TLogger = &tLogger{0, nil, map[*regexp.Regexp]int{}, false}
+	TLogger = &tLogger{errors: map[*regexp.Regexp]int{}}
 	vLevel := os.Getenv("GRPC_GO_LOG_VERBOSITY_LEVEL")
 	if vl, err := strconv.Atoi(vLevel); err == nil {
 		TLogger.v = vl
 	}
 }
 
+// getStackFrame gets, from the stack byte string, the appropriate stack frame.
 func getStackFrame(stack []byte, frame int) (string, error) {
 	s := strings.Split(string(stack), "\n")
 	if frame >= (len(s)-1)/2 {
@@ -69,6 +73,7 @@ func getStackFrame(stack []byte, frame int) (string, error) {
 	return fmt.Sprintf("%v:", split[len(split)-1]), nil
 }
 
+// log logs the message with the specified parameters to the tLogger.
 func (g *tLogger) log(ltype logType, format string, args ...interface{}) {
 	s := debug.Stack()
 	prefix, err := getStackFrame(s, callingFrame)
@@ -116,6 +121,8 @@ func (g *tLogger) Update(t *testing.T) {
 		g.initialized = true
 	}
 	g.t = t
+	g.m.Lock()
+	defer g.m.Unlock()
 	g.errors = map[*regexp.Regexp]int{}
 }
 
@@ -135,11 +142,15 @@ func (g *tLogger) ExpectErrorN(expr string, n int) {
 		g.t.Error(err)
 		return
 	}
+	g.m.Lock()
+	defer g.m.Unlock()
 	g.errors[re] += n
 }
 
 // EndTest checks if expected errors were not encountered.
 func (g *tLogger) EndTest(t *testing.T) {
+	g.m.Lock()
+	defer g.m.Unlock()
 	for re, count := range g.errors {
 		if count > 0 {
 			t.Errorf("Expected error '%v' not encountered", re.String())
@@ -148,7 +159,10 @@ func (g *tLogger) EndTest(t *testing.T) {
 	g.errors = map[*regexp.Regexp]int{}
 }
 
+// expected determines if the error string is protected or not.
 func (g *tLogger) expected(s string) bool {
+	g.m.Lock()
+	defer g.m.Unlock()
 	for re, count := range g.errors {
 		if re.FindStringIndex(s) != nil {
 			g.errors[re]--
